### PR TITLE
Clasrel

### PR DESCRIPTION
**Pull request by Silvicek**
Adds Argus yes/no task and data set

### CODE REVIEW

File: tasks/yesno.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
+
```
Comment (pasky) [inline tasks/yesno.py]: Interesting.  I have expected that you'd just modify the hypev task appropriately.  Is there a reason to keep the old hypev task in addition to the yesno task?  Is "yesno" a fitting name?  What about the old hypev argus data?

File: tasks/yesno.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+from keras.layers.embeddings import Embedding
+from keras.models import Graph
+from pysts.clasrel_layers import Reshape_, WeightedMean
+import pysts.nlp as nlp
+
+from . import AbstractTask
+
+
+class Container:
+    """Container for merging questions together."""
```
Comment (pasky) [inline tasks/yesno.py]: Isn't it about merging atexts (etexts) for the same question, rather than merging questions?

File: tasks/yesno.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+                print("save")
+
+        return (gr, y, vocab)
+
+    def build_model(self, module_prep_model, do_compile=True):
+
+        model = build_model(self.emb, self.vocab, module_prep_model, self.c)
+
+        # for lname in self.c['fix_layers']:  # whats this??
+        #     model.nodes[lname].trainable = False
```
Comment (pasky) [inline tasks/yesno.py]: What's the reason to comment this out and not to support fixing some layers in place?

File: tasks/yesno.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        return tuple(res)
+
+    def res_columns(self, mres, pfx=' '):
+        """ Produce README-format markdown table row piece summarizing
+        important statistics """
+        return('%s%.6f    |%s%.6f |%s%.6f |%s%.6f'
+               % (pfx, mres[self.trainf]['Loss'],
+                  pfx, mres[self.valf]['Precision'],
+                  pfx, mres[self.testf].get('Loss', np.nan),
+                  pfx, mres[self.testf].get('Precision', np.nan)))
```
Comment (pasky) [inline tasks/yesno.py]: Interesting choice.  Why train + test reports loss (what does it tell us compared to precision?) while val reports precision?

File: tasks/yesno.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+        N = glove.N + nlp.flagsdim
+    else:
+        N = glove.N
+
+    model.add_shared_node(name='embdrop', inputs=['e0', 'e1'], outputs=['e0_', 'e1_'],
+                          layer=Dropout(dropout, input_shape=(N,)))
+
+    return N
+
+def build_model(glove, vocab, module_prep_model, c):
```
Comment (pasky) [inline tasks/yesno.py]: Any reason why this isn't a method of the task?

File: tasks/yesno.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    model.add_node(name='scoreS1', input=mlp_ptscorer(model, final_outputs, c['Ddim'], N, c['l2reg'], pfx='S1_'),
+                   layer=Activation(oact))
+    model.add_node(name='scoreS2', input=mlp_ptscorer(model, final_outputs, c['Ddim'], N, c['l2reg'], pfx='S2_'),
+                   layer=Activation(oact))
+
+
+def embedding(model, glove, vocab, s0pad, s1pad, dropout, trainable=False,
+              add_flags=False):
+    """ Sts embedding layer, without creating inputs. """
+    # TODO: add switch to original -> no code duplication
```
Comment (pasky) [inline tasks/yesno.py]: Please do. :)

File: tasks/yesno.py (hunk 1 above)
Comment (pasky) [inline tasks/yesno.py]: Actually, it might make sense to have two models that are similar, but one uses the hypothesis aggregation like yesno while the other just averages pair predictions, for comparison purposes.  But they must at least be able to process the same dataset, so could we use the hypev format and load_hypev()?

File: tasks/yesno.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+                   merge_mode='concat', concat_axis=-1)
+    # ===================== mean of class over rel
+    model.add_node(WeightedMean(w_dim=rnn_dim,
+                                q_dim=rnn_dim,
+                                max_sentences=max_sentences), name='weighted_mean', input='c_r')
+    model.add_output(name='score', input='weighted_mean')
+    return model
+
+def mlp_ptscorer(model, inputs, Ddim, N, l2reg, pfx='out', sum_mode='sum'):
+    """ Element-wise features from the pair fed to an MLP. """
```
Comment (pasky) [inline tasks/yesno.py]: Please make ptscorer configurable.  Can we make normal mlp_ptscorer work like this in case Ddim == 0?

File: tasks/yesno.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+    # model.add_node(name=pfx+'hdn', inputs=[pfx+'sum', pfx+'mul'], merge_mode='concat',
+    #                layer=Dense(output_dim=int(N*Ddim), W_regularizer=l2(l2reg), activation='sigmoid'))
+    # model.add_node(name=pfx+'mlp', input=pfx+'hdn',
+    #                layer=Dense(output_dim=1, W_regularizer=l2(l2reg)))
+    model.add_node(name=pfx+'mlp', inputs=[pfx+'sum', pfx+'mul'], merge_mode='concat',
+                   layer=Dense(output_dim=1, W_regularizer=l2(l2reg), activation='linear'))
+    return pfx+'mlp'
+
+from models.pysts.kerasts.blocks import LambdaMerge
+def absdiff_merge(model, layers):
```
Comment (pasky) [inline tasks/yesno.py]: Why do we need own copy of this?

File: pysts/clasrel_layers.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
+from keras.layers.wrappers import TimeDistributed as td
```
Comment (pasky) [inline pysts/clasrel_layers.py]: Should this file be in pysts/ instead of kerasts/ ?

File: pysts/clasrel_layers.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+from keras import activations, initializations
+import keras.constraints
+import keras.regularizers
+from keras.layers.core import MaskedLayer, Layer, TimeDistributedDense, TimeDistributedMerge, Activation
+import keras.backend as K
+import theano.tensor as T
+import numpy as np
+
+def relu(x):
+    return K.switch(x > 0, x + 0.01, 0.01)
```
Comment (pasky) [inline pysts/clasrel_layers.py]: Huh, keras doesn't have relu, or what's going on?

**Comment by pasky** [comment]
Thanks! It's great work and I'm very happy that it's finally coming back to dataset-sts.

I think it needs a little bit more work though, I have expected this would share more with the hypev task; at the very least, it should share the data format.  Maybe we can rename the tasks to have more similarity in the name too.

I also don't really like the copied over functions, imho the task structure is slightly messy :) - more in the comments.  I'll be also happy if you could add some more comments, esp. descriptions of modules and classes, but this is lower priority.

File: tasks/yesno.py (hunk 1 above)
Comment (Silvicek) [inline tasks/yesno.py]: The hypev task learns on separate samples and the averaging is done only when evaluating. Load_hypev is therefore unusable in yesno. I completely forgot about hypev and will delete it..
The old argus data are still usable in yesno (but should we use it?)

File: tasks/yesno.py (hunk 2 above)
Comment (Silvicek) [inline tasks/yesno.py]: yup :(

File: tasks/yesno.py (hunk 3 above)
Comment (Silvicek) [inline tasks/yesno.py]: wasnt sure what this does, will uncomment

File: tasks/yesno.py (hunk 4 above)
Comment (Silvicek) [inline tasks/yesno.py]: my bad

File: tasks/yesno.py (hunk 1 above)
Comment (pasky) [inline tasks/yesno.py]: On Sun, Apr 24, 2016 at 02:23:57AM -0700, Silvicek wrote:

> > @@ -0,0 +1,301 @@
> > +
> 
> The hypev task learns on separate samples and the averaging is done only when evaluating. Load_hypev is therefore unusable in yesno. In my opinion the hypev task is outdated..

I don't understand why does the averaging imply that load_hypev() is
unusable, can you elaborate?

What we need to show in any papers we publish is that our method has
merit compared to a simpler averaging baseline.

Comment (Silvicek) [inline tasks/yesno.py]: load_hypev is unusable because the training only works with pairs <question>-<sentence> and not with <question>-<bunch of sentences>. The precision hypev shows is only averaged over questions after the training. 
The correct averaging baseline should use the same end-to-end approach as the current clasrel implementation only with simple average at the end.

Comment (pasky) [inline tasks/yesno.py]: First, I'm not sure what does that have to do with loading data.
load_anssel doesn't aggregate the second sentences either, does it?

Second, if "hypev" is obsolete, why not name this "hypev"?  I'm not sure
how appropriate "yesno" is e.g. for entrance exam datasets.  It's just
nitpicking, of course, but I don't understand the name change.

File: pysts/clasrel_layers.py (hunk 10 above)
Comment (Silvicek) [inline pysts/clasrel_layers.py]: Before the introduction of stabilizing constant in the clasrel denominator, this was a necessary relu adjustment, now its outdated.

File: tasks/yesno.py (hunk 1 above)
Comment (Silvicek) [inline tasks/yesno.py]: Sure i'll rename it :P

**Comment by Silvicek** [comment]
Ok, i addressed most of the issues. It was necessary to create additional ptscorer because the use of the hidden layer in mlp_ptscorer caused val performance decrease.

File: pysts/vocab.py [hunk 11]
```diff
@@ -56,7 +56,12 @@ def embmatrix(self, emb):
             try:
                 embedding_weights[index, :] = emb.g[word]
             except KeyError:
-                embedding_weights[index, :] = np.random.uniform(-0.25, 0.25, emb.N)  # 0.25 is embedding SD
+                if index == 0:
+                    embedding_weights[index, :] = np.zeros(emb.N)
+                else:
+                    embedding_weights[index, :] = np.random.uniform(-0.25, 0.25, emb.N)  # 0.25 is embedding SD
+
+
```
Comment (pasky) [inline pysts/vocab.py:64]: Adventurous change!  However, it'll impact all other models too.  I have started a sanity check evaluation on yoda dataset with just this change to confirm if we'll need to rebenchmark everything again or it's just a small change.

File: pysts/kerasts/blocks.py [hunk 12]
```diff
@@ -185,6 +186,20 @@ def mlp_ptscorer(model, inputs, Ddim, N, l2reg, pfx='out', sum_mode='sum', extra
     return pfx+'mlp'
 
 
+def mulsum_ptscorer(model, inputs, Ddim, N, l2reg, pfx='out', sum_mode='sum'):
+    """ Element-wise features from the pair fed to a concat of sum and multiplication of inputs.
+    """
```
Comment (pasky) [inline pysts/kerasts/blocks.py:191]: Hmm, I still don't quite understand why we can't just special-case mlp_ptscorer for Ddim=0.   But we don't need to dwell on this, I can try to change that myself.

File: tasks/hypev.py [hunk 13]
```diff
@@ -3,95 +3,249 @@
...
         """ Produce README-format markdown table row piece summarizing
         important statistics """
-        return('%s%.6f |%s%.6f |%s%.6f'
-               % (pfx, mres[self.trainf]['QAccuracy'],
-                  pfx, mres[self.valf]['QAccuracy'],
-                  pfx, mres[self.testf].get('QAccuracy', np.nan)))
+        return('%s%.6f |%s%.6f |%s%.6f '
+               % (pfx, mres[self.trainf]['Precision'],
+                  pfx, mres[self.valf]['Precision'],
+                  pfx, mres[self.testf].get('Precision', np.nan)))
```
Comment (pasky) [inline tasks/hypev.py]: So, is it precision or accuracy? I thought .evaluate() reports accuracy, and that's probably what we want too.

File: pysts/vocab.py (hunk 11 above)
Comment (Silvicek) [inline pysts/vocab.py:64]: Come to think about it, because of the implemented masking, reverting this should not make any impact on the clasrel results. However using nonzero vectors for nonexistent words is very weird.

Comment (pasky) [inline pysts/vocab.py:64]: It looks clear to change, no impact. I agree it's nicer.

**Comment by pasky** [comment]
(I think we are clear to merge, except I'd love to clear up the Precision - QAccuracy change logic.)

**Comment by pasky** [comment]
thx
